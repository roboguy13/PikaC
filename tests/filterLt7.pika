%generate filterLt7

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

filterLt7 : Sll -> Sll;
filterLt7 Nil := Nil [Sll];
filterLt7 (Cons a as) | a < 7 := filterLt7 as;
filterLt7 (Cons a as) | not (a < 7) := Cons a (filterLt7 as) [Sll];

%test "filter1" Sll:
  filterLt7 (Cons 1 (Cons 2 (Cons 10 (Cons 20 (Nil [Sll]) [Sll]) [Sll]) [Sll]) [Sll]);

