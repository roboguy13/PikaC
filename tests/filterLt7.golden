void filterLt7(loc x, loc x1);
void filterLt7(loc x, loc x1) {
 loc x2 = READ_LOC(x1, 0);
 if (x < 7) {
  if (x == 0) {
   WRITE_LOC(x1, 0, 0);
  } else {
   loc h = READ_LOC(x, 0);
   loc n = READ_LOC(x, 1);
   if (h < 7) {
    filterLt7(n, x1);
   } else {
    if (n < 7) {
     filterLt7(n, x1);
     WRITE_LOC(x, 0, n);
    } else {
     filterLt7(n, x1); WRITE_LOC(x, 0, x);
    }
   }
  }
 } else {
  if (x1 < 7) {
   loc h = READ_LOC(x, 0);
   loc n = READ_LOC(x, 1);
   if (h < 7) {
    filterLt7(n, x1);
   } else {
    if (n < 7) {
     filterLt7(n, x1);
     WRITE_LOC(x, 0, n);
    } else {
     if ((n <= x) && (x <= n)) {
      filterLt7(n, x1);
      free( x );
     } else {
      filterLt7(n, x1); WRITE_LOC(x, 0, x1);
     }
    }
   }
  } else {
   if (x2 < 7) {
    loc h = READ_LOC(x, 0);
    loc n = READ_LOC(x, 1);
    if (h < 7) {
     filterLt7(n, x1);
    } else {
     if (n < 7) {
      filterLt7(n, x1);
      WRITE_LOC(x, 0, n);
     } else {
      if ((n <= x) && (x <= n)) {
       filterLt7(n, x1);
       free( x );
      } else {
       filterLt7(n, x1); WRITE_LOC(x, 0, x2);
      }
     }
    }
   } else {
    loc h = READ_LOC(x, 0);
    loc n = READ_LOC(x, 1);
    if (h < 7) {
     filterLt7(n, x1);
    } else {
     if (n < 7) {
      filterLt7(n, x1);
      WRITE_LOC(x, 0, n);
     } else {
      if ((n <= x) && (x <= n)) {
       filterLt7(n, x1);
       free( x );
      } else {
       filterLt7(n, x1); WRITE_LOC(x, 0, 0);
      }
     }
    }
   }
  }
 }
}